From 645eabfd61d07ccf91964fd49460e9bc4dc6fc90 Mon Sep 17 00:00:00 2001
From: Jan Petrous <jan.petrous@nxp.com>
Date: Tue, 2 Nov 2021 13:55:44 +0100
Subject: [PATCH 16/19] driver:net:pfeng: detect PFE version

For S32G3 bringup we need to differentiate on silicon version
on runtime.

Issue: ALB-8061
Upstream-Status: Pending 

Signed-off-by: Jan Petrous <jan.petrous@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/pfeng/hw_abi/pfeng_hw.h |  1 +
 drivers/net/pfeng/pfeng_hw.c        | 23 +++++++++++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/drivers/net/pfeng/hw_abi/pfeng_hw.h b/drivers/net/pfeng/hw_abi/pfeng_hw.h
index 0bf4422fd0..f433a952e1 100644
--- a/drivers/net/pfeng/hw_abi/pfeng_hw.h
+++ b/drivers/net/pfeng/hw_abi/pfeng_hw.h
@@ -74,6 +74,7 @@ struct pfe_platform {
 	u32 hif_chnl_count; /* Number of HIF channels */
 	u32 emac_count;	    /* Number of EMAC blocks */
 	u8  emac_mdio_div;  /* Divider for mdio clk */
+	u8  on_g3;          /* True if running on S32G3 */
 	u32 gpi_total;	    /* Number of GPI/ETGPI/HGPI blocks */
 	u32 gpi_count;	    /* Number of GPI blocks */
 	u32 etgpi_count;	    /* Number of ETGPI blocks */
diff --git a/drivers/net/pfeng/pfeng_hw.c b/drivers/net/pfeng/pfeng_hw.c
index 19d80f49ec..e3a34521a1 100644
--- a/drivers/net/pfeng/pfeng_hw.c
+++ b/drivers/net/pfeng/pfeng_hw.c
@@ -28,6 +28,10 @@
 
 #include "pfe_cbus.h"
 
+#define WSP_VERSION			(0)
+#define WSP_VERSION_SILICON_G2		0x00050300
+#define WSP_VERSION_SILICON_G3		0x00000101
+
 static struct pfe_platform pfe;
 
 static const struct pfe_ct_hif_tx_hdr header[3U] = {
@@ -1905,6 +1909,25 @@ int pfeng_hw_init(struct pfe_platform_config *config)
 		goto exit;
 	}
 
+	/* Check version */
+	addr = (u32 *)((uint64_t)pfe.cbus_baseaddr +
+				 CBUS_GLOBAL_CSR_BASE_ADDR + WSP_VERSION);
+	val = readl(addr);
+	if (val == WSP_VERSION_SILICON_G2) {
+		/* S32G2 */
+		pr_debug("PFE: S32G2 detected\n");
+		pfe.on_g3 = false;
+	} else if (val == WSP_VERSION_SILICON_G3) {
+		/* S32G3 */
+		pr_debug("PFE: S32G3 detected\n");
+		pfe.on_g3 = true;
+	} else {
+		/* Unknown */
+		pr_err("PFE: Silicon HW version is unknown: 0x%x\n", val);
+		pfe.on_g3 = false;
+	}
+
+	/* Init LMEM */
 	addr = (u32 *)(void *)((uint64_t)pfe.cbus_baseaddr +
 				    CBUS_LMEM_BASE_ADDR);
 	pr_debug("Initializing LMEM (%d bytes)\n", (int)CBUS_LMEM_SIZE);
-- 
2.17.1

